src/Culling/ray.ts

```diff
@@ -671,22 +692,31 @@ Scene.prototype._internalPick = function (rayFunction: (world: Matrix) => Ray, p
             continue;
         }
 
-        var world = mesh.skeleton && mesh.skeleton.overrideMesh ? mesh.skeleton.overrideMesh.getWorldMatrix() : mesh.getWorldMatrix();
-        var ray = rayFunction(world);
+        let world = mesh.skeleton && mesh.skeleton.overrideMesh ? mesh.skeleton.overrideMesh.getWorldMatrix() : mesh.getWorldMatrix();
 
-        var result = mesh.intersects(ray, fastCheck, trianglePredicate, onlyBoundingInfo);
-        if (!result || !result.hit) {
-            continue;
-        }
+        let result = this._internalPickForMesh(pickingInfo, rayFunction, mesh, world, fastCheck, onlyBoundingInfo, trianglePredicate);
 
-        if (!fastCheck && pickingInfo != null && result.distance >= pickingInfo.distance) {
-            continue;
+        if (result) {
+            pickingInfo = result;
+
+            if (fastCheck) {
+                return pickingInfo;
+            }
         }
 

```

Comment:
If the mesh has thin instances (and picking is enabled), we should not check for intersection with the mesh world matrix but only with the thin instance matrices. The mesh itself is not rendered when it has thin instances: if one wants to render the mesh, `thinInstanceAddSelf` should be called (and then a matrix will be added to the matrix buffer).

Reply:
Good point!